# Digit Classification using MNIST Dataset 
 

### Load Dataset


In [22]:
from tensorflow.keras.datasets import mnist
import keras
from keras import layers
import numpy as np
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

In [18]:
print(f"Training data shape: {train_images.shape}\nTest data shape: {test_images.shape}\n\nTrain Labels: {np.unique(train_labels)}")

Training data shape: (60000, 784)
Test data shape: (10000, 784)

Train Labels: [0 1 2 3 4 5 6 7 8 9]


### Preprocess


In [30]:
# Convert 3D data to 2D as NN input will be just an array of vectors
train_images = train_images.reshape((60000, 28 * 28))           # (60000, 784)
test_images = test_images.reshape((10000, 28 * 28))

# Convert grayscale pixel values into values in range 0-1. 
# (As neural network weights are usually small, so less computation will be needed to update them)
train_images = train_images.astype('float32') / 255
test_images  = test_images.astype('float32') / 255

### Build a simple sequential model


In [31]:
model = keras.Sequential([
    layers.Dense(512, input_shape=(784,), activation='relu'),
    layers.Dense(10, activation='softmax')
])

/home/sabyr/.local/lib/python3.10/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [32]:
model.compile(
    optimizer='rmsprop',                    # can deal with the issue of local minima 
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

### Train the Model

In [33]:
model.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 4s 7ms/step - accuracy: 0.8747 - loss: 0.4306
Epoch 2/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9667 - loss: 0.1154
Epoch 3/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9786 - loss: 0.0729
Epoch 4/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9847 - loss: 0.0494
Epoch 5/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9889 - loss: 0.0371


### Performance Evaluation on Test Data

In [35]:
model.evaluate(test_images, test_labels)

313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9779 - loss: 0.0729


[0.0628848671913147, 0.98089998960495]

97% accuracy with such a simple DNN is actually worthwhile